### PR TITLE
FXiOS:1483 Sync and open tab consolidation

### PR DESCRIPTION
**Pull request by nbhasin2**
Basic Changes
- Added support for showing sync tabs, open tabs along with bookmarks and history in awesomebar (URL)

General Issues
- Fixed issue with empty tab title in recently closed list under library 

Refactoring
Panels updated: Downloads, History, Recently Closed, Remote Tabs

a) TwoLineHeaderFooterView
- Updated Panel with new TwoLineHeaderFooterView and TwoLineImageOverlayCell 

b) TwoLineImageOverlayCell
- Removed older two line cell and replaced it with TwoLineImageOverlayCell
- Supports RTL 
- Supports Dark mode

c)  Refactored OneLineCell to be used in SearchViewController
- Since it is used in Library view it will will also match part of new designs

d) Added missing colors Greay75 with alpha 60%

Improvement: Add support for accessibility when changing text size for TwoLineHeaderFooterView and TwoLineImageOverlayCell

Screen Shot 2021-04-05 at 5 18 29 PM

### CODE REVIEW

**Comment by nbhasin2** [comment]
Opened PR for review, might push minor cleanup items / nits that I see needs fixing up. 

Since this PR touches core functionality of our Awesome Bar (URL) its important to take a deeper look at functional items and let me know if I missed anything. 

I have also touched all / most of the library panels as I refactored two line cell so please try to test those as well and bring to my attention an issues you find. 

Thanks

**Comment by isabelrios** [comment]
@nbhasin2 I was taking a look at the test failures and looks like these three tests are failing because the swipe to delete item is not working in this PR for the history or download panel (I guess it will be the same for all panels): 
-DownloadFilesTests.testDownloadFile()
-DownloadFilesTests.testToastButtonToGoToDownloads()
-HistoryTests.testDeleteHistoryEntryBySwiping()
Is that expected? sorry if the work is not finished yet, but was worried that you were blocked on this PR due to test failures.
Thanks!

**Comment by nbhasin2** [comment]
@isabelrios Thanks for checking, even I saw those tests failing however locally both items work. I am still working on this PR but it might be because of refactoring I did. Will take a look at fixing these tests

<table>
  <tr>
    <td>History</td>
     <td>Bookmark</td>
     <td>Downloads</td>
     <td>Reading List</td>
  </tr>
  <tr>
    <td>Screen Shot 2021-04-06 at 10 48 15 AM </td>
    <td>Screen Shot 2021-04-06 at 10 48 01 AM</td>
    <td>Screen Shot 2021-04-06 at 10 54 12 AM</td>
    <td>Screen Shot 2021-04-06 at 10 55 10 AM</td>
  </tr>
 </table>

**Comment by isabelrios** [comment]
> @isabelrios Thanks for checking, even I saw those tests failing however locally both items work. I am still working on this PR but it might be because of refactoring I did. Will take a look at fixing these tests

oh okay, sorry, I tried within your branch running the tests but also using the app and was not able to see tha Delete menu after swiping...

**Comment by nbhasin2** [comment]
@electricRGB Tagging you for an early review while I look at some test failures 

Also, I added new design for sync tab here as it'll be moved to our tab tray so thought it'll be easier to do it here

**Comment by nbhasin2** [comment]
@irios Even after enabling accessibility element they I still can't make it swipe using swipeleft() method in XCUITest. Although everything works fine on local run so maybe you know a method or something that I might be missing.

**Review by electricRGB — APPROVED** [review]
Approving for now. Seems good to me, pending fixing the tests I guess. :)

File: Client/Frontend/Widgets/TwoLineCellAndFooter.swift [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            make.height.width.equalTo(20)
+            make.trailing.equalTo(leftImageView).offset(7)
+            make.bottom.equalTo(leftImageView).offset(7)
+        }
+
+        selectedBackgroundView = selectedView
+        applyTheme()
+    }
+    
+    func applyTheme() {
```
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Client/Frontend/Widgets/TwoLineCellAndFooter.swift:116]: No hard rule, both are fine. `Themable` can be added to extension (depends on scenario) or to a class itself as long we are consistent with overall file structure.

File: Client/Frontend/Library/DownloadsPanel.swift [hunk 2]
```diff
@@ -81,7 +81,7 @@ class DownloadsPanel: UIViewController, UITableViewDelegate, UITableViewDataSour
 
         tableView.delegate = self
         tableView.dataSource = self
-        tableView.register(TwoLineTableViewCell.self, forCellReuseIdentifier: "TwoLineTableViewCell")
+        tableView.register(TwoLineImageOverlayCell.self, forCellReuseIdentifier: "TwoLineImageOverlayCell")
```
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Client/Frontend/Library/DownloadsPanel.swift:84]: Good catch, will refactor this :)

File: Client/Frontend/Browser/SearchViewController.swift [hunk 3]
```diff
@@ -297,7 +318,70 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
             self.view.layoutIfNeeded()
         })
     }
+    
+    fileprivate func getCachedTabs() {
+        assert(Thread.isMainThread)
+        // Short circuit if the user is not logged in
+        guard profile.hasSyncableAccount() else {
```
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Client/Frontend/Browser/SearchViewController.swift]: Fixed :)

**Comment by nbhasin2** [comment]
https://user-images.githubusercontent.com/8919439/113956267-38252800-97eb-11eb-814f-ce9414ec4216.mov

Woohoo got that swipe working in a XCUItest, only had to refactor major part of code in https://github.com/mozilla-mobile/firefox-ios/pull/8268/commits/c5abedd4404c3fbf5e367ab79d3e5b2589d6fa17 😅 

Lets see if this works in Bitrise 🤞🏼

**Comment by nbhasin2** [comment]
@isabelrios Finally got it working, the `StackView` somehow was taking all the user gestures. If I turn off `userinteractionenabled` then everything worked but instead I just got rid of stackview and went with good old `container view` route. It was little more work but this is full proof solution that will definitely not have any issues. 

Tests Passing: 
- https://app.bitrise.io/build/54d306cfa93e00dc#?tab=log
- https://github.com/mozilla-mobile/firefox-ios/runs/2292842486

Commit: https://github.com/mozilla-mobile/firefox-ios/pull/8268/commits/c5abedd4404c3fbf5e367ab79d3e5b2589d6fa17

**Comment by nbhasin2** [comment]
Thank you so much for your help @isabelrios 🙏🏼

**Comment by isabelrios** [comment]
> Lets see if this works in Bitrise 🤞🏼

Yay! Bitrise is ✅ for this PR now!

**Pull request merged**